File: eclipse-plugin/src/main/java/xyz/devglobe/eclipse/core/CoreClient.java
package xyz.devglobe.eclipse.core;

import java.io.*;
import java.util.concurrent.TimeUnit;

/**
 * JSON-line IPC client for the devglobe-core subprocess.
 * Mirrors the CoreClient from the JetBrains plugin.
 */
public class CoreClient {

    // ── Callbacks ───────────────────────────────────────────────────────

    public interface Callbacks {
        void onReady(boolean configured);
        void onNotConfigured();
        void onInvalidApiKey();
        void onHeartbeatOk(int todaySeconds, String language);
        void onOffline();
        void onOnline();
        void onStatusOk();
        void onStatusError(String message);
        void onError(String message);
        void onProcessDied();
    }

    // ── Fields ───────────────────────────────────────────────────────────

    private Process process;
    private BufferedWriter writer;
    private Thread readerThread;
    private volatile boolean running = false;
    private final Callbacks callbacks;

    public CoreClient(Callbacks callbacks) {
        this.callbacks = callbacks;
    }

    // ── Lifecycle ────────────────────────────────────────────────────────

    public synchronized boolean start(String binaryPath) {
        if (running) return true;
        try {
            ProcessBuilder pb = new ProcessBuilder(binaryPath, "daemon");
            pb.redirectErrorStream(false);

            process = pb.start();
            writer = new BufferedWriter(new OutputStreamWriter(process.getOutputStream()));

            readerThread = new Thread(this::readLoop, "DevGlobe-CoreReader");
            readerThread.setDaemon(true);
            readerThread.start();

            running = true;
            return true;
        } catch (IOException e) {
            DevGlobePlugin.log("Failed to start core: " + e.getMessage());
            callbacks.onError("Failed to start core: " + e.getMessage());
            return false;
        }
    }

    public synchronized void stop() {
        try {
            if (writer != null) {
                sendShutdown();
                writer.flush();
            }
        } catch (IOException ignored) {}
        running = false;
        try {
            if (process != null) process.waitFor(1, TimeUnit.SECONDS);
        } catch (InterruptedException ignored) {
            Thread.currentThread().interrupt();
        }
        try {
            if (writer != null) writer.close();
        } catch (IOException ignored) {}
        if (process != null) process.destroyForcibly();
        if (readerThread != null) readerThread.interrupt();
        process = null;
        writer = null;
        readerThread = null;
    }

    public boolean isRunning() {
        return running && process != null && process.isAlive();
    }

    // ── Send methods ─────────────────────────────────────────────────────

    public synchronized void sendInit(String pluginVersion, String editor) {
        send("init", "plugin_version", pluginVersion, "editor", editor);
    }

    public synchronized void sendActivity(String file, String language) {
        if (language != null && !language.isEmpty()) {
            send("activity", "file", file, "language", language);
        } else {
            send("activity", "file", file);
        }
    }

    public synchronized void sendSetStatus(String message) {
        send("set_status", "message", message);
    }

    public synchronized void sendPause() {
        send("pause");
    }

    public synchronized void sendResume() {
        send("resume");
    }

    public synchronized void sendShutdown() {
        send("shutdown");
    }

    // ── Internal ─────────────────────────────────────────────────────────

    private void send(String method, Object... keyValues) {
        if (!running || writer == null) return;
        try {
            StringBuilder sb = new StringBuilder();
            sb.append("{\"method\":\"").append(method).append("\"");
            if (keyValues != null && keyValues.length >= 2) {
                sb.append(",\"params\":{");
                for (int i = 0; i < keyValues.length; i += 2) {
                    if (i > 0) sb.append(",");
                    sb.append("\"").append(keyValues[i]).append("\":");
                    Object val = keyValues[i + 1];
                    if (val instanceof Number) {
                        sb.append(val);
                    } else {
                        sb.append("\"").append(escapeJson(val.toString())).append("\"");
                    }
                }
                sb.append("}");
            }
            sb.append("}\n");
            writer.write(sb.toString());
            writer.flush();
        } catch (IOException e) {
            DevGlobePlugin.log("Failed to send to core: " + e.getMessage());
        }
    }

    private void readLoop() {
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(process.getInputStream()))) {
            String line;
            while (running && (line = reader.readLine()) != null) {
                handleLine(line.trim());
            }
        } catch (IOException ignored) {
        } finally {
            running = false;
            callbacks.onProcessDied();
        }
    }

    private void handleLine(String line) {
        if (line.isEmpty()) return;
        try {
            // Minimal JSON parsing — no external dependency
            String event = extractString(line, "event");
            if (event == null) return;

            switch (event) {
                case "ready": {
                    boolean configured = extractBool(line, "configured");
                    callbacks.onReady(configured);
                    break;
                }
                case "not_configured":
                    callbacks.onNotConfigured();
                    break;
                case "invalid_api_key":
                    callbacks.onInvalidApiKey();
                    break;
                case "heartbeat_ok": {
                    int secs = extractInt(line, "today_seconds");
                    String lang = extractString(line, "language");
                    callbacks.onHeartbeatOk(secs, lang);
                    break;
                }
                case "offline":
                    callbacks.onOffline();
                    break;
                case "online":
                    callbacks.onOnline();
                    break;
                case "status_ok":
                    callbacks.onStatusOk();
                    break;
                case "status_error": {
                    String msg = extractString(line, "message");
                    callbacks.onStatusError(msg != null ? msg : "Unknown status error");
                    break;
                }
                default:
                    DevGlobePlugin.log("Unknown event from core: " + event);
            }
        } catch (Exception e) {
            DevGlobePlugin.log("Error parsing core message: " + e.getMessage());
        }
    }

    // ── Minimal JSON helpers ─────────────────────────────────────────────

    private static String extractString(String json, String key) {
        String needle = "\"" + key + "\":\"";
        int start = json.indexOf(needle);
        if (start < 0) {
            // try without quotes (number, bool, or null literal)
            needle = "\"" + key + "\":";
            start = json.indexOf(needle);
            if (start < 0) return null;
            start += needle.length();
            int end = json.indexOf(",", start);
            if (end < 0) end = json.indexOf("}", start);
            if (end < 0) end = json.length();
            String raw = json.substring(start, end).trim();
            // JSON null literal → Java null
            if ("null".equals(raw)) return null;
            return raw;
        }
        start += needle.length();
        int end = json.indexOf("\"", start);
        if (end < 0) return null;
        return json.substring(start, end);
    }

    private static int extractInt(String json, String key) {
        String val = extractString(json, key);
        if (val == null) return 0;
        try { return Integer.parseInt(val); } catch (NumberFormatException e) { return 0; }
    }

    private static boolean extractBool(String json, String key) {
        String val = extractString(json, key);
        return "true".equalsIgnoreCase(val);
    }

    private static String escapeJson(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"")
                .replace("\n", "\\n").replace("\r", "\\r").replace("\t", "\\t");
    }
}

File: jetbrains-plugin/src/main/kotlin/xyz/devglobe/plugin/core/CoreClient.kt
package xyz.devglobe.plugin.core

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.intellij.openapi.Disposable
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.ApplicationInfo
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

class CoreClient(private val binaryPath: String) : Disposable {

    private val LOG = Logger.getInstance(CoreClient::class.java)

    private var process: Process? = null
    private var writer: BufferedWriter? = null
    private var readerThread: Thread? = null

    var onReady: ((Boolean) -> Unit)? = null
    var onNotConfigured: (() -> Unit)? = null
    var onHeartbeatOk: ((Int, String?) -> Unit)? = null
    var onOffline: (() -> Unit)? = null
    var onOnline: (() -> Unit)? = null
    var onStatusOk: (() -> Unit)? = null
    var onStatusError: ((String) -> Unit)? = null
    var onError: ((String) -> Unit)? = null
    var onProcessDied: (() -> Unit)? = null

    fun start(): Boolean {
        if (process != null) return true
        try {
            val proc = ProcessBuilder(binaryPath, "daemon").redirectErrorStream(false).start()
            process = proc
            writer = BufferedWriter(OutputStreamWriter(proc.outputStream, Charsets.UTF_8))

            readerThread = Thread({ readLoop(proc) }, "DevGlobe-CoreReader").apply {
                isDaemon = true
                start()
            }

            LOG.info("Core daemon started: $binaryPath")
            return true
        } catch (e: Exception) {
            LOG.error("Failed to start core daemon: ${e.message}")
            onError?.invoke("Failed to start devglobe-core: ${e.message}")
            return false
        }
    }

    private fun readLoop(proc: Process) {
        try {
            BufferedReader(InputStreamReader(proc.inputStream, Charsets.UTF_8)).useLines { lines ->
                lines.forEach(::handleLine)
            }
        } catch (_: Exception) {
            // stream closed — fall through to onProcessDied
        }
        ApplicationManager.getApplication().invokeLater { onProcessDied?.invoke() }
    }

    fun isAlive(): Boolean = process?.isAlive == true

    fun sendInit(pluginVersion: String) {
        val params = JsonObject().apply {
            addProperty("plugin_version", pluginVersion)
            addProperty("editor", detectEditor())
        }
        send("init", params)
    }

    fun sendActivity(filePath: String, language: String?) {
        val params = JsonObject().apply {
            addProperty("file", filePath)
            language?.let { addProperty("language", it) }
        }
        send("activity", params)
    }

    fun sendSetStatus(message: String) {
        val params = JsonObject().apply {
            addProperty("message", message)
        }
        send("set_status", params)
    }

    fun sendPause() = send("pause", null)
    fun sendResume() = send("resume", null)
    fun sendShutdown() = send("shutdown", null)

    private fun send(method: String, params: JsonObject?) {
        try {
            val msg = JsonObject().apply {
                addProperty("method", method)
                if (params != null) add("params", params)
            }
            writer?.write(msg.toString())
            writer?.newLine()
            writer?.flush()
        } catch (e: Exception) {
            LOG.warn("Failed to send to core: ${e.message}")
        }
    }

    private fun handleLine(line: String) {
        try {
            val obj = JsonParser.parseString(line).asJsonObject
            val event = obj.get("event")?.asString ?: return
            val data = obj.getAsJsonObject("data")

            ApplicationManager.getApplication().invokeLater {
                when (event) {
                    "ready" -> {
                        val configured = data?.get("configured")?.asBoolean ?: false
                        onReady?.invoke(configured)
                    }
                    "not_configured" -> onNotConfigured?.invoke()
                    "heartbeat_ok" -> {
                        val todaySeconds = data?.get("today_seconds")?.asInt ?: 0
                        val language = data?.get("language")?.takeIf { !it.isJsonNull }?.asString
                        onHeartbeatOk?.invoke(todaySeconds, language)
                    }
                    "offline" -> onOffline?.invoke()
                    "online" -> onOnline?.invoke()
                    "status_ok" -> onStatusOk?.invoke()
                    "status_error" -> onStatusError?.invoke(data?.get("message")?.asString ?: "")
                }
            }
        } catch (e: Exception) {
            LOG.warn("Failed to parse core event: ${e.message}")
        }
    }

    private fun detectEditor(): String {
        val productCode = ApplicationInfo.getInstance().build.productCode
        return when (productCode) {
            "IC", "IU" -> "intellij"
            "WS" -> "webstorm"
            "PY", "PC", "PE" -> "pycharm"
            "GO" -> "goland"
            "PS" -> "phpstorm"
            "RM" -> "rubymine"
            "CL" -> "clion"
            "RD" -> "rider"
            "DB" -> "datagrip"
            "AI" -> "android-studio"
            "RR" -> "rustrover"
            else -> "jetbrains"
        }
    }

    override fun dispose() {
        try { sendShutdown() } catch (_: Exception) { }
        process?.destroyForcibly()
        process = null
        writer = null
    }
}
